File: 프로그래머스/2/42842. 카펫/카펫.py
def get_divisors(n):
    divisors = []
    for i in range(1, int(n**(1/2))+1):
        if n % i == 0:
            divisors.append((n//i, i)) # 약수쌍 저장
    return divisors

def solution(brown, yellow):  
    total = brown + yellow # 전체 격자의 수
    divisors = get_divisors(total)
    
    for w, h in divisors:
        if w < 3 or h < 3: # 가로, 세로 모두 3 이상이어야 가능
            continue
        if (w-2) * (h-2) == yellow:
            return [w, h]

File: 프로그래머스/2/42842. 카펫/카펫.java
import java.util.*;

class Solution {
    static List<int[]> divisors = new ArrayList<>();
    
    public int[] solution(int brown, int yellow) {
        int[] answer = new int[2];
        
        int total = brown + yellow;
        getDivisors(total);
        
        for (int[] divisor: divisors) {
            int w = divisor[0], h = divisor[1];
            if (w < 3 || h < 3) {
                continue;
            } else {
                if ((w-2) * (h-2) == yellow) {
                    answer[0] = w;
                    answer[1] = h;
                }
            }
        }
        return answer;
    }
    
    private void getDivisors(int n) {
        for (int i = 1; i <= Math.sqrt(n); i++) {
            if (n % i == 0) {
                divisors.add(new int[]{n / i, i});
            }
        }
    }
}